### ISSUE TITLE
The usage of wxgtk3-dev breaks compatibility to other software

### ISSUE THREAD

**Issue by MartinX3** [moneymanagerex/moneymanagerex#4255]
MMEX version:
 - [x] 1.5.x
 - [ ] 1.6.x

Note: bug reporters are expected to have verified the bug still exists
either in the last stable version of MMEX or on updated development code
(master branch).

Operating System:
 - [ ] Windows
 - [ ] Mac OSX
 - [x] Linux 

# Description of the bug

We are forced to use the development version of wxWidgets, which breaks compatibility of other software in linux depending on the official wxWidgets version.

# Reproduction

Is the bug reproducible? 
 - [x] Always 
 - [ ] Randomly 
 - [ ] Happened only once

Reproduction steps:

1. Install the package of wxWidgets 3.1.x
2. install
   1. filezilla 
   2. wxMaxima
   3. mediainfo-gui
   4. Audacity
5. start them

Expected result:
Working apps.

Actual result:
Graphical issues.
Crash on startup.
As example `wxMaxima` dies with the message `wxmaxima: error while loading shared libraries: libwx_gtk3u_richtext-3.0.so.0: cannot open shared object file: No such file or directory`.

On the other side moneyManagerEx crashes under wxWidgets 3.0.x.

Maybe it would be a good longtherm goal to get rid of wxWidgets or fix the crash reason in 3.0.x.

I also don't know why the last release of wxWidgets is from 2021-04-14.
They are so slow, having around 1900 error tickets and making the CI builds of mmex a burden.

# Additional information

N/A

**Comment by renato-mmex** [comment]
MartinX3 write:
_They are so slow, having around 1900 error tickets and making the CI builds of mmex a burden._

do not agree with this statement, because many of the reported bugs are personal wishes of the users to MMEX and not bugs. Unfortunately, the programming capacity is very limited, which is unavoidable in a project like MMEX, because there is nothing to earn. In addition, because of the Russian crisis, the Internet is no longer available for Nikolay and causes further problems with the development. 
It is easy to make demands on a product that is developed and maintained by volunteers in their free time. I think MMEX has made a lot of progress in the last few years. However, it is very difficult to maintain MMEX in all the different operating systems. I would be very happy if more professionals would join the project.

**Comment by MartinX3** [comment]
@renato-mmex I try to help the project as well.
I maintain the arch linux package and added commits to maintain a buildable source code.

However, if you look at the wxWidgets source code repository, they are slow and have 1900 open issue tickets. Just take a look there.
There is also no logical reason why their latest development release is 1 year old. Their latest stable release is 2 years old.
wxWidget has 2 companies and 282 contributors supporting the project, also 4200 github stars and and an amount of project forks so large that the github webinterface doesn't work correctly.

I said wxWidgets is slow, not mmex.
(Maybe my message got lost in my text wall.)

mmex is nice, but also shares the disadvantages of other c(++).
The older the project code gets, the more compiler warnings we see.
Also c(++) is hard to maintain without a very good documentation, refactoring and code patterns and even if you try your best, the c(++) code will be hard to understand and to maintain. Especially if the project dependencies age as well or become end of life but still are used in the project.

But what I want to say is, that users installing mmex in their linux distribution via package manager will break other packages in their system using wxWidgets 3.0.x. And the package repositories won't upgrade to 3.1.x since it's a development release and not a stable release. So the developers for as example filezilla or wxMaxima won't update their dependencies to wxWidgets 3.1.x until the wxWidgets project releases a new stable version to upgrade to.

**Comment by tactilis** [comment]
@renato-mmex 
> do not agree with this statement, because many of the reported bugs are personal wishes of the users to MMEX and not bugs. 

He wasn't referring to MMEX

**Comment by PMaff** [comment]
FileZilla Client Version 3.28.0 is working on my Ubuntu 18.04 machine.
Linked against:
  wxWidgets:      3.0.3
  SQLite:         3.22.0
  GnuTLS:         3.5.18

I compiled and  installed wxWidget 3.1.5 and use that in mmex.
Never used wxMaxima .

**Comment by MartinX3** [comment]
@PMaff
Arch Linux here
1. wxWidgets 3.0.x and 3.1.x are conflicting so I can't install them together
2. starting filezilla results into `$ filezilla 
filezilla: error while loading shared libraries: libwx_gtk3u_aui-3.0.so.0: cannot open shared object file: No such file or directory`

How do you start it without bundling the wxWidgets 3.0.x with it?

**Comment by PMaff** [comment]
Did you try to compile  wxWidgets 3.1.5 ?
What does configure say when starting to do so?
I get:
```
wxWidgets-3.1.5/gtk_build$ ../configure 

config.status: executing wx-config commands

Configured wxWidgets 3.1.5 for `x86_64-pc-linux-gnu'

  Which GUI toolkit should wxWidgets use?                 GTK+ 3 with support for GTK+ printing libnotify
  Should wxWidgets be compiled into single library?       no
  Should wxWidgets be linked as a shared library?         yes
  Should wxWidgets support Unicode?                       yes (using wchar_t)
  What level of wxWidgets compatibility should be enabled?
                                       wxWidgets 2.8      no
                                       wxWidgets 3.0      yes
  Which libraries should wxWidgets use?
                                       STL                no
                                       jpeg               sys
                                       png                sys
                                       regex              builtin
                                       tiff               sys
                                       lzma               yes
                                       zlib               sys
                                       expat              sys
                                       libmspack          no
                                       sdl                no
```
and:
```
wx-config --list

    Default config is gtk3-unicode-3.1

  Default config will be used for output

  Alternate matches:
    gtk2-unicode-3.0
    gtk2-unicode-3.1
```

**Comment by PMaff** [comment]
filezilla is the package that I got from the Ubuntu repos.
Nothing changed there.
I still have libwxgtk3.0-gtk3-0v5 installed on Ubuntu 18.04 :
https://packages.ubuntu.com/bionic/amd64/libwxgtk3.0-gtk3-0v5/filelist

**Comment by MartinX3** [comment]
In arch linux the packages `wxgtk3` and `wxgtk-dev` are in conflict with each other since both overwrite the file `/usr/bin/wx-config-gtk3`

https://archlinux.org/packages/extra/x86_64/wxgtk3/
https://aur.archlinux.org/packages/wxgtk3-dev

**Comment by PMaff** [comment]
I do not have any of the wxgtk v3.1.5 packages from Ubuntu since these are not official for Ubuntu 18.04 LTS.
I have to compile wxWidget 3.1.5 myself from the sources.
Finally installing them as root with "make install" which obviously installs them parallel to v3.0.x .
There is no "package" which I have to install in another step.

As you see, I do not know the way, how Arch Linux installs binaries, libraries that were compiled by yourself.
Maybe "make install" as root acts different on Arch Linux and Ubuntu.

**Comment by MartinX3** [comment]
`make install` should overwrite your `/usr/bin/wx-config-gtk3` as well.
If `/usr/bin/wx-config-gtk3` doesn't exist in your wxgtk3 package from the ubuntu repo, they don't use the default files of wxgtk3.

The Arch Linux AUR installs binaries like the normal repos. It just has an extra step before to generate the package file.
In Ubuntu it's a `.deb` file, in arch linux it's a `.zstd` file.

**Comment by PMaff** [comment]
which seems to be the case: 
```
ll /usr/bin/wx-config-gtk3
ls: cannot access '/usr/bin/wx-config-gtk3': No such file or directory
```

```
/usr/local/bin/wx-config -> /usr/local/lib/wx/config/gtk3-unicode-3.1
/usr/bin/wx-config -> /etc/alternatives/wx-config
/etc/alternatives/wx-config -> /usr/lib/x86_64-linux-gnu/wx/config/gtk2-unicode-3.0
```

`/usr/lib/x86_64-linux-gnu/wx/config/gtk2-unicode-3.0`
is the older exec.
And this is the newer one:
`/usr/local/lib/wx/config/gtk3-unicode-3.1`

Yes, that step with a .deb file after doing "make install" is missing on Ubuntu.

**Comment by MartinX3** [comment]
for v3.1.5 it is

```$ whereis wx-config
wx-config: /usr/bin/wx-config

$ ls -lha /usr/bin/wx-config
lrwxrwxrwx 1 root root 35 18. Mär 11:43 /usr/bin/wx-config -> /usr/lib/wx/config/gtk3-unicode-3.1
```

**Comment by PMaff** [comment]
Do you have some gtk2-unicode-3.0 in parallel?

It looks like Ubuntu and Opensuse can install the wxWidget 3.1.5 in parallel to the  3.0.3.
Question is, if Arch Linux can do that too.

**Comment by PMaff** [comment]
Hmm, I am not an Arch Linux pro but I see that you use
https://aur.archlinux.org/packages/wxgtk3-dev
according to your above answer.
AUR is the user-contributed according to https://wiki.archlinux.org/title/Arch_Build_System .
Does this mean that you did not compile wxWidget 3.1.5 from scratch from the sources?

**Comment by MartinX3** [comment]
Since both packages overwrite `/usr/bin/wx-config` it is not possible and both wxWidgets need to stay in conflict with each other.
And since Arch Linux is not Debian or one of its children there is no `update-alternatives`.

I compile it via the AUR installer, but it is installed as a system package, like MMEX is installed as a system package.
That's how you should use Linux.
Some self compiled custom stuff in custom paths isn't supported.

**Comment by PMaff** [comment]
So you cannot compile from sources and simply install that on Arch Linux?

How do you test software where you write the sources and that you do not want to put into AUR at first?

**Comment by MartinX3** [comment]
Why should I garbage my system with custom stuff as an enduser?
I use MMEX to maintain my finances and not to develop it.
Why should I use Linux if it is forbidden to just install and use software?

And as I said before, the AUR script does a `make install`

**Comment by PMaff** [comment]
I see your point but since I am coming from the developer corner I do want to compile things myself and install them
for testing without any official AUR stuff.
If such software is ok, then I think about AUR or similar official repos.
Of course I am responsible then to keep track what is on the system if I have selfcompiled stuff (gftp, wxWidget, mmex).

I do fear that the AUR scripts are doing something differentl than Ubuntu and Suse (so 2 different Linux distros)
when installing such source compiled stuff.
To be honest: I am happy that Ubuntu and Suse are not interfering with some additional scripts when I compile from sources
;-)

AUR scripts therefore seem to be missing the way how wxWidget 3.1.5 installs on Ubuntu AND Suse,

**Comment by MartinX3** [comment]
I don't know why you're religious about Ubuntu, Suse and self compiling stuff.
A normal user just wants to use a linux distribution without tampering around with custom compiled stuff.

About the AUR-Script here: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=wxgtk-dev
It just compiles the source, use the `make install` to create the proper files needed to install it as a system package and installs it as a system package.
It's the same how a normal repo package gets installed, except that you generate the package file on your local machine first in AUR.

There is nothing what's wrong and need to be corrected in the AUR buildscript.
3.0.x and 3.1.x are overwriting each others `/usr/bin/wx-config-gtk3`

Btw. gftp is a package here in our arch linux repos. Thanks to Arch Linux and AUR we don't need to play around with PPA's, custom repos and self compiled custom placed stuff on our system. Especially if the Ubuntu packages are outdated again.

**Comment by PMaff** [comment]
Maybe this helps:
"Currently the Arch Linux packages mix the runtime files with the development-only files because a wxWidgets installation also installs them all together...
With this fix the runtimes of both 3.0.x and 3.1.x can now be installed together on a system."
https://github.com/andriybyelikov/archlinux-wxwidgets-packages

**Comment by MartinX3** [comment]
> Script to switch between wxWidgets configurations (only needed for development)

That's an unofficial custom hack for developers.
Not for endusers.

**Comment by PMaff** [comment]
Since wxWidget 3.1.5 is a developer version this would mean that there is no solution in Arch Linux for "endusers"
and all Arch Linux "endusers" have to wait, till wxWidget >= 3.1.5 is official and supported in the official repos in Arch Linux.
Sorry.
@whalley : I do think that this ticket can be closed as "Won't fix" for mmex.

**Comment by MartinX3** [comment]
3.1.x will never be official, since it's marked as "development" by its devs.
3.2.x should be the next stable version, maybe next year. I don't know how fast/slow they are.
Especially since 3.0.5 and 3.1.5 are old releases.

Just if other people having problems because of this dependency, they won't use mmex everywhere where dev and stable versions of a package can't coexist since they overwrite the same file in the Linux system.

**Comment by PMaff** [comment]
@renato-mmex :
>  because of the Russian crisis, the Internet is no longer available for Nikolay and causes further problems with the development. 

Is he completely cut off from internet? :-/

**Comment by MartinX3** [comment]
Shit :(

**Comment by vomikan** [comment]
> @renato-mmex :
> 
> > because of the Russian crisis, the Internet is no longer available for Nikolay and causes further problems with the development.
> 
> Is he completely cut off from internet? :-/

Everything is fine with the internet for me.
American social networks such as Facebook and Instagram are blocked.

**Comment by whalley** [comment]
> @whalley : are there plans to use wxWidget 3.0.X and wxWidget 3.1.5 in parallel in the configure scripts (and sources)?

To be honest I have no intention of doing anything to support 3.0.x, I want to move forward. I equally appreciate the issue this may cause for Linux end users but am going to have to leave it to the ingenuity of Linux developers to help solve this.

**Comment by MartinX3** [comment]
To avoid misunderstandings:
It's not important to stay at an old software version.
The main problem here is, that the wxWidget developers never declared 3.1.x stable, so it can't be pushed to the package repos of linux distributions as stable version.

**Comment by PMaff** [comment]
@whalley : I understand this.
It would be lot of work to rewrite the sources for 3.0.X and 3.1.5 in parallel, many #ifdefs in the source code, quite a change of the configure files.

**Comment by PMaff** [comment]
> To avoid misunderstandings: It's not wrong to cut the support of old libraries. The main problem here is, that the wxWidget developers never declared 3.1.x stable, so it can't be pushed to the package repos of linux distributions as a stable version for default usage.

The reason for not putting this to "stable" is named on the website:
"Please notice that while 3.1.5 is officially a “development” version because it is not fully compatible with the “stable” 3.0.x, the list of backwards incompatible changes is very short, so you shouldn’t have any problems updating to this version from 3.0.x in practice, and you’re encouraged to use this release, including in production."
https://www.wxwidgets.org/

If it is on the pure words, then of course it is a "development" version...also for production. :-)
Since one deb- and one rpm-ecosystem are able to separate "production" wxWidget 3.0.X and "development for production" version 3.1.5 I do not see the problem on mmex side.

**Comment by MartinX3** [comment]
> > To avoid misunderstandings: It's not wrong to cut the support of old libraries. The main problem here is, that the wxWidget developers never declared 3.1.x stable, so it can't be pushed to the package repos of linux distributions as a stable version for default usage.
> 
> The reason for not putting this to "stable" is named on the website: "Please notice that while 3.1.5 is officially a “development” version because it is not fully compatible with the “stable” 3.0.x, the list of backwards incompatible changes is very short, so you shouldn’t have any problems updating to this version from 3.0.x in practice, and you’re encouraged to use this release, including in production." [wxwidgets.org](https://www.wxwidgets.org/)

They break the semantic versioning standard by adding API incompatible changes and only raising the minor version of the library.
That's a bad practice.
And also their 3.1.5 version isn't enrolled as ready production package in the repos.

> If it is on the pure words, then of course it is a "development" version...also for production. :-) Since one deb- and one rpm-ecosystem are able to separate "production" wxWidget 3.0.X and "development for production" version 3.1.5 I do not see the problem on mmex side.

In which official repo are the wxGTK 3.1.x and 3.0.x packages at the same time?
Are you referring to hacked custom repos you need to add to linux?

It's not the case in debian, fedora or ubuntu.

**Comment by PMaff** [comment]
Of course they could have named it 4.X.
But that does not change the main problem here.

Arch User Repository (AUR) is offering packages from users.
The guy who offers the 3.1.5 development(!) version of wxWidget in AUR should care about not overwriting anything of the production 3.0.X .
It is possible since the `make install` from wxWidget config does not overwrite 3.0.X per se.
Neither in Ubuntu (deb ecosystem) nor in Suse (rpm ecosystem).

**Comment by PMaff** [comment]
It is not victim blaming, it is just a description, how it is working on other ecosystems (Ubuntu and Suse).
`make install` should then also overwrite my 3.0.x installation on Ubuntu and Suse, if it was the error on the installation scripts
of wxWidget 3.1.5 .

It is not overwriting anything on Ubuntu or Suse. 
Therefore - logically- there must be something different with the scripts from AUR package of 3.1.5.
This might be also the reason for 
https://github.com/andriybyelikov/archlinux-wxwidgets-packages 

boost was just another example, which installs newer (compiled from source) versions (1.78.0) parallel to official repos (1.66.0 ) from e.g. Suse:
https://software.opensuse.org/package/boost

How do you test a newer version of e.g. boost libs e.g. 1.78.0, if you want to keep the old 1.66.0 on Arch?
Do you have to uninstall 1.66.0 and then install 1.78.0 , try, then uninstall 1.78.0 and then reinstall 1.66.0?

**Comment by MartinX3** [comment]
These two packages conflict with each other, because they generate at least one file with the same name and path if you execute `make install`.

If you upgrade a package in a Linux OS you overwrite it.
If you want to test libs before everyone else you change your update channel from stable to testing.

According to your questions I strongly advise you to read documentations about linux packaging for developers.

**Comment by PMaff** [comment]
Be patient with me.
I do not know Arch Linux and it's "update channels".

In my example when testing a newer version I do not need to upgrade the official repo package.
I simply `configure/make/make install` from the source and get the version to test in
/usr/local/bin/
while some official version from the official repos still is available in
/usr/bin/
For example:
older official MPlayer 1.3.0
3817184 Jän 30  2018 /usr/bin/mplayer
and self compiled newer MPlayer 1.4-7
20209976 Jän  2 23:55 /usr/local/bin/mplayer

Probably there is still a misunderstanding: when I compile wxWidget or other software from source and do a `make install` on Ubuntu or Suse, I do not make a package for my installation. 
`make install` scripts from wxWidget are not doing this on Ubuntu and Suse.
If I want to make packages for repos, I have to do additional steps to produce a rpm or deb ,
which are not part of the `make install` from wxWidget.

So in fact I do have three possibilities for testing a new software e.g. in Suse:
1. get the update for the official repo package
2. get the user made newer community package
3. compile from source and install via `make install` in /usr/local/bin/ parallel to an official version.

I did 3. with wxWidget 3.1.5, whereas you did 2. on Arch as far as I understand.

**Comment by MartinX3** [comment]
I'm sorry if I sounded harsh.

Every linux distribution has its testing channels, since they also need the community to test and report bugs.
We are all volunteers and users at the same time somehow.
In arch linux the channels are
- staging
- testing
- stable
where staging possible break your system.

In Ubuntu the place where updates are tested is called `proposed`.
But Ubuntu also has a rolling release mode where you distro hop to their newest release point currently in development.

Debian has
- unstable
- testing
- next-stable

Fedora has the `updates-testing` repository.

Only AUR does a `make install` on the enduser computer.
A repo maintainer does it in his CI, creates the package, uploads it to the repo and we can install it.
AUR is just a way to introduce a new package which maybe added to the official repos if enough people vote for it.

The 3.1.x and 3.0.x package contents going into
`/usr/bin/`, `/usr/lib/` , `/usr/share/`

AUR packages are just repo packages created on the fly on the local computer of the person who installs it.

It's more hard for people using as example ubuntu.
They need to add an external ppa for the wxgtk3 3.1.x version since the switch to 3.1.x in mmex.
And if mmex was a part of the official repos, it got removed then.

**Comment by PMaff** [comment]
That mechanism is all clear.
The point 3) that I mentioned is without any deb/rpm (or alike) package and without any repo maintainer.
I do not know, which method puts a self compiled lib into
`/usr/local/lib/`
or a self compiled binary into `/usr/local/bin`
without using AUR for ARCH linux?

**Comment by MartinX3** [comment]
Arch Packages are system packages where you do the packaging step of the package maintainer on your own computer.
So they need to follow the same rules as packages in the standard repos.

**Comment by PMaff** [comment]
These rules seem to prohibit installation of development packages for testing in parallel to "production" ones.
That is the reason, why I am doing 3) (see above) and not 2) .

**Comment by MartinX3** [comment]
That's the reason mmex got removed from the official repos and is forced to stay in the AUR.

And there normal people won't be able to install it via package manager terminal/gui if the dependencies aren't installed by the package management.

**Comment by PMaff** [comment]
As long as you use AUR and as long as you do not compile/make install development version yourself parallel to production version, this will be the case.
It is therefore just a matter, who does the `make install` : you as the one who then installs it parallel to the production 3.0.X
(which is my 3) in the above)
or
the developer/maintainer of some AUR repo package, who does not do a `make install` in parrallel location.
(which is the 2) in the above).

But both methods 2) and 3) are completely valid.

**Comment by MartinX3** [comment]
The AUR package maintainer writes the CI script and the package installation executes it.
There I can either install it on my system (default) or generate a .zst package file (like your .dep or .rpm)

And the manual of wxWidgets says we need to `make install` and also follow the guidelines of the arch linux package maintaining.

**Comment by PMaff** [comment]
Hmm, I only read `make install` here:
"Building the libraries"
Might be interesting for you:
"Note that installing the library is not required and it can also be used from the build directory by simply specifying the full path to wx-config script located in it when building your application."

**Comment by MartinX3** [comment]
So I should bloat the mmex package by adding third party packages to it, instead of require it in the package system via the package meta data?

That's no custom "build it yourself thing". It needs to act as a package you install via a linux package manager from a repo.
Also mmex is not the only package in the linux distribution eco system depending on a wxwidgets package as a dependency.

**Comment by PMaff** [comment]
It is not "bloating".
It is simply the separation of production stuff from development stuff.
See
"Then you may install the library and its header files under /usr/local/include/wx and /usr/local/lib respectively"
there: https://docs.wxwidgets.org/3.1.5/plat_gtk_install.html

I do only see the wxWidget compile instructions there, nothing special for ARCH Linux, and therefore I still assume, that the AUR scripts are overwriting production wxWidget 3.0.X .
Maybe ARCH Linux is  not suited to build development libraries yourself by simply installing development libs like 3.1.5 or 3.1.6 in parallel to production libraries 3.0.X?
Of course it would be more nice to have the 3.1.6 as production libs but for the time being this is no problem as long as you do not overwrite production libs when installing any development libs (in a package or with `make install` according to
https://docs.wxwidgets.org/3.1.5/plat_gtk_install.html
when you compile yourself as it is done in Ubuntu/Suse).

**Comment by MartinX3** [comment]
We install mmex via the package manager to use it, not to develop it.

Arch Linux does nothing special.
It uses `make install` and follows the https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard to put the files into the right path.
So it puts the files into `/usr` like `/usr/lib`.
It does not use `/usr/local/...` since it would violate with the intended use case.

I see there is a wxgtk package variation putting it in `/opt`.
I look into updating my mmex aur package. Maybe it will work.

**Comment by MartinX3** [comment]
I tried now to break compatibility with the standard wxWidgets package and switched to a custom package installing them at `/opt/wxgth-dev/`
Sadly the fresh compiled mmex can't finx the wxWidgets libs.

I use
` cmake -DCMAKE_BUILD_TYPE=Release -DwxWidgets_CONFIG_EXECUTABLE=/opt/wxgtk-dev/bin/wx-config-gtk3 -DwxWidgets_LIB_DIR=/opt/wxgtk-dev/lib/ .`

And I see the error message
```
CMake Warning:
  Manually-specified variables were not used by the project:

    wxWidgets_LIB_DIR
```

**Comment by vomikan** [comment]
>Sadly the fresh compiled mmex can't finx the wxWidgets libs.

May be export path=/opt/wxgtk-dev/lib/ && mmex

**Comment by MartinX3** [comment]
Is there a way to bake the path into the mmex code?
The normal enduser just clicks on the desktop icon or the entry in his taskbar menu and don't write bash variables in the terminal.

**Comment by vomikan** [comment]
> Is there a way to bake the path into the mmex code?
> The normal enduser just clicks on the desktop icon or the entry in his taskbar menu and don't write bash variables in the terminal.

First, at the level of advanced users, we need to find a solution.

**Comment by vomikan** [comment]
FYI 
https://stackoverflow.com/questions/31092593/how-can-i-use-cmake-to-both-build-wxwidgets-on-demand-and-link-with-it

**Comment by MartinX3** [comment]
```$ path=/opt/wxgtk-dev/lib/ && ./mmex 
./mmex: error while loading shared libraries: libwx_gtk3u_core-3.1.so.6: cannot open shared object file: No such file or directory
```

Does the last link require to modify C++ code?
I'm not a C++ developer.

**Comment by whalley** [comment]
Is it not LD_LIBRARY_PATH you need to set? 

`$ LD_LIBRARY_PATH=/opt/wxgtk-dev/lib/ && ./mmex `

What does `ldd -v mmex` report ?

**Comment by MartinX3** [comment]
```bash
 LD_LIBRARY_PATH=/opt/wxgtk-dev/lib/ && ./mmex 
./mmex: error while loading shared libraries: libwx_gtk3u_core-3.1.so.6: cannot open shared object file: No such file or directory
```
and the shorted report
```bash
$ ldd -v mmex
	libwx_gtk3u_core-3.1.so.6 => /usr/lib/libwx_gtk3u_core-3.1.so.6 (0x00007f3fa14b2000)
	libwx_gtk3u_qa-3.1.so.6 => /usr/lib/libwx_gtk3u_qa-3.1.so.6 (0x00007f3fa147f000)
	libwx_gtk3u_html-3.1.so.6 => /usr/lib/libwx_gtk3u_html-3.1.so.6 (0x00007f3fa1397000)
	libwx_baseu_xml-3.1.so.6 => /usr/lib/libwx_baseu_xml-3.1.so.6 (0x00007f3fa1381000)
	libwx_gtk3u_aui-3.1.so.6 => /usr/lib/libwx_gtk3u_aui-3.1.so.6 (0x00007f3fa12d7000)
	libwx_gtk3u_stc-3.1.so.6 => /usr/lib/libwx_gtk3u_stc-3.1.so.6 (0x00007f3fa1048000)
	libwx_gtk3u_webview-3.1.so.6 => /usr/lib/libwx_gtk3u_webview-3.1.so.6 (0x00007f3fa1016000)
	libwx_baseu-3.1.so.6 => /usr/lib/libwx_baseu-3.1.so.6 (0x00007f3fa0cf2000)

	Version information:
	./mmex:
		libwx_baseu_xml-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_baseu_xml-3.1.so.6
		libwx_gtk3u_aui-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_gtk3u_aui-3.1.so.6
		libwx_gtk3u_qa-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_gtk3u_qa-3.1.so.6
		libwx_gtk3u_html-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_gtk3u_html-3.1.so.6
		libwx_gtk3u_webview-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_gtk3u_webview-3.1.so.6
		libwx_gtk3u_stc-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_gtk3u_stc-3.1.so.6
		libwx_baseu-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_baseu-3.1.so.6
		libwx_gtk3u_core-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_gtk3u_core-3.1.so.6
	/usr/lib/libwx_gtk3u_core-3.1.so.6:
		libwx_baseu-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_baseu-3.1.so.6
	/usr/lib/libwx_gtk3u_qa-3.1.so.6:
		libwx_baseu_xml-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_baseu_xml-3.1.so.6
		libwx_gtk3u_core-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_gtk3u_core-3.1.so.6
		libwx_baseu-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_baseu-3.1.so.6
	/usr/lib/libwx_gtk3u_html-3.1.so.6:
		libwx_gtk3u_core-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_gtk3u_core-3.1.so.6
		libwx_baseu-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_baseu-3.1.so.6
	/usr/lib/libwx_baseu_xml-3.1.so.6:
		libwx_baseu-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_baseu-3.1.so.6
	/usr/lib/libwx_gtk3u_aui-3.1.so.6:
		libwx_baseu-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_baseu-3.1.so.6
		libwx_gtk3u_core-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_gtk3u_core-3.1.so.6
	/usr/lib/libwx_gtk3u_stc-3.1.so.6:
		libwx_gtk3u_core-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_gtk3u_core-3.1.so.6
		libwx_baseu-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_baseu-3.1.so.6
	/usr/lib/libwx_gtk3u_webview-3.1.so.6:
		libwx_gtk3u_core-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_gtk3u_core-3.1.so.6
		libwx_baseu-3.1.so.6 (WXU_3.1) => /usr/lib/libwx_baseu-3.1.so.6
	/usr/lib/libwx_baseu-3.1.so.6:
	/usr/lib/libwebkit2gtk-4.0.so.37:

```

**Comment by PMaff** [comment]
Your mmex seems to be linked with some older libs in /usr/lib/
and not in /opt/wxgtk-dev/lib/ .
See your CMake Warning .
By "switched to a custom package installing " you mean the way of installing AUR packages on ARCH linx
or
the normal way of doing the `make install` yourself?

**Comment by MartinX3** [comment]
`make install` and putting it in `/opt/wxgtk-dev/`

The cmake warning only happens because I manually added `wxWidgets_LIB_DIR` to the mmex compilation.
By default mmex looks into the wxWidgets standard path.

**Comment by moritzhoewer** [comment]
@MartinX3 are you sure you need the `wxWidgets_LIB_DIR`? During my testing to find the webview bug (#4269), I at some point built mmex against a local wxWidgets in my home folder (not installed as a system package) just by changing the `wxWidgets_CONFIG_EXECUTABLE` to `/home/.../wxWidgets/wx-config`. This worked as expected and `ldd` also showed me that mmex was linking the libs in my home folder (and not the ones from the system). And when I re-buit that local wxWidgets with a different configuration, mmex would no longer start (as would be expected).

I am also on arch linux and I used your PKGBUILD as a base (changing line 38 as described above to use my local wxWidgets). The resulting package installed and worked normally - until I "broke" its local library dependencies as described above

**Comment by MartinX3** [comment]
I tried the same with only the config executable without success before but I could try it again later. :(

**Comment by moritzhoewer** [comment]
Maybe the `wx-config` itself doesn't know it's supposed to be in `/opt/wxgt-dev` and outputs the wrong paths?

**Comment by PMaff** [comment]
@MartinX3 I got mmex running on Manjaro, does this help you for ARCH?
https://github.com/moneymanagerex/moneymanagerex/issues/4154#issuecomment-1104560001

**Comment by MartinX3** [comment]
It gives me
```bash
$ /opt/wxgtk-dev/lib/wx/config/gtk3-unicode-3.1 --libs
-L/opt/wxgtk-dev/lib -pthread   -lwx_gtk3u_xrc-3.1 -lwx_gtk3u_html-3.1 -lwx_gtk3u_qa-3.1 -lwx_gtk3u_core-3.1 -lwx_baseu_xml-3.1 -lwx_baseu_net-3.1 -lwx_baseu-3.1
```

**Comment by PMaff** [comment]
Just for fun I compiled wxmaxima-Version-22.03.0.tar.gz from scratch.
Guidelines are here:
https://github.com/wxMaxima-developers/wxmaxima/blob/main/Compiling.md

When starting wxmaxima I do not get error messages about wxWidget and any missing wxWidget libraries :

```
[pmaff@virtualmanjaro build-wxm]$ wxmaxima 

(wxmaxima:4621): Gtk-WARNING **: 22:13:17.332: Failed to measure available space: Error getting filesystem info for /media/sf_GemeinsamerOrdner: Permission denied

(wxmaxima:4621): Gtk-WARNING **: 22:13:22.651: Failed to measure available space: Error getting filesystem info for /media/sf_GemeinsamerOrdner: Permission denied
```

**Comment by MartinX3** [comment]
It depends if you compiled wxmaxima with the wxwidget package in the repos (3.0) or if you compile it with the development package from aur (3.1).

so you can't use filezilla, wxMaxima, mediainfo-gui and mmex together.

**Comment by PMaff** [comment]
I compiled wxMaxima with the self compiled wxWidget, that I also use for mmex.
Nothing from AUR and wxWidget was never installed via packages.

**Comment by MartinX3** [comment]
`pacman -S filezilla`

```bash
$ filezilla 
filezilla: error while loading shared libraries: libwx_gtk3u_aui-3.0.so.0: cannot open shared object file: No such file or directory
```

**Comment by PMaff** [comment]
I just tried on the Manjaro virtual machine with Plasma and Wayland that is currently running: no such message when 
running filezilla from terminal. filezilla comes up normal.
Difference: I never installed any wxWidget from any package but only (!) compiled version 3.1.6 from source.

**Comment by PMaff** [comment]
But why does a filezilla, that has been installed via 
`pacman -S filezilla`
not complain about shared libraries, when wxWidget 3.1.6 was compiled from source and not installed via pacman?

**Comment by MartinX3** [comment]
Because you didn't install mmex via pacman before.

**Comment by PMaff** [comment]
Are you installing an older wxWidget (older than -say- 3.1.5)version together with mmex via pamac?

**Comment by MartinX3** [comment]
3.0.5 is the latest stable release of wxWidgets
Apps in the repo use the wxWidgets 3.0.5 from the repo

mmex gets installed via aur with the wxwidgets 3.1.6 aur package

**Comment by PMaff** [comment]
Strange...so every program other than mmex that uses wxWidget libraries should use the newly installed wxWidget 3.1.6 as it is done in my case with filezilla on Manjaro (without ever installing a 3.0.5 before).
What happens if you uninstall 3.0.5 before installing mmex with wxWidget 3.1.6 via aur?
Does filezilla then run with 3.1.6?

**Comment by MartinX3** [comment]
You need to recompile filezilla yourself.
Currently I don't have 3.0.5 installed, since I migrated my wxWidgets applications away to solutions not using wxWidgets.

The package `filezilla` installs the wxWidgets
https://archlinux.org/packages/community/x86_64/filezilla/

**Comment by PMaff** [comment]
That's what is on the Manjaro system now:
```
ll /usr/lib/wx                                                                      
total 12
drwxr-xr-x 3 root root 4096 21. Mai 19:13 3.0
drwxr-xr-x 2 root root 4096 21. Mai 19:13 config
drwxr-xr-x 3 root root 4096 21. Mai 19:13 include

ll /usr/local/lib/wx                                                                   
total 12
drwxr-xr-x 3 root root 4096 21. Mai 19:10 3.1.6
drwxr-xr-x 2 root root 4096 21. Mai 19:10 config
drwxr-xr-x 3 root root 4096 21. Mai 19:10 include

wx-config --version                                                           
3.1.6
```

I also installed audacity now which uses wxWidget 3.0.5 .
No complaint about 
"error while loading shared libraries" or alike.

Maybe the trick is to first compile the wxWidget 3.1.6 from sources, install it with simple `make install` 
without pamac, aur or alike 
and then install packages that use 3.0.5 ?

**Comment by MartinX3** [comment]
you installed mmex, filezilla and audacity via the package managers in arch linux?

**Comment by PMaff** [comment]
I compiled wxWidget 3.1.6 from sources and installed with `make install`
then
I compiled mmex from sources and installed with `make install`
then
I installed filezilla and audacity via pamac.

**Comment by MartinX3** [comment]
Endusers install Software from packages.

I don't compile the wxWidgets manually.
The needed package is added as a dependency in the metadata of the mmex pkgbuild file.

https://aur.archlinux.org/packages/moneymanagerex-git
https://aur.archlinux.org/packages/wxgtk3-dev
https://aur.archlinux.org/packages/wxgtk3-dev-opt

**Comment by PMaff** [comment]
So the author of  https://aur.archlinux.org/packages/wxgtk3-dev has compiled it from
"https://github.com/wxWidgets/wxWidgets/releases/download/v3.1.6/wxWidgets-3.1.6.tar.bz2" as I can see
"Sources (1)" there?
Where does this package install wxWidget 3.1.6 ?
In /usr/lib/wx ?

**Comment by MartinX3** [comment]
Accordig to the pkgbuild it's `/usr/` like the 3.0.5 package from the repos.

**Comment by PMaff** [comment]
Of course that won't work.
But why does the author of https://aur.archlinux.org/packages/wxgtk3-dev and alike install 3.1.6 as if it was a official version?
If all the other packages expect 3.0.5  there, then putting 3.1.6 there _must_ result in havoc.
That is the reason, why when compiling 3.1.6 this is installed in /usr/local/lib/wx and not in the place, where 3.0.5 is expected
by other packages.

Others btw. seem to have the same problem as you:
"After installed it if launch filezilla I have:

filezilla: error while loading shared libraries: libwx_gtk3u_aui-3.0.so.0: cannot open shared object file: No such file or directory

I guest must be changed "provided" in wxgtk3-dev or filezilla dependecies"
https://aur.archlinux.org/packages/wxgtk3-dev

This looks like rather the author of the https://aur.archlinux.org/packages/wxgtk3-dev package made a mistake rather than
the wxWidget guys or mmex makes a mistake.
Same for https://aur.archlinux.org/packages/wxgtk3-dev-opt .
After installing these with the same location as 3.0.5 then you would have to recompile/repack all those programs that expect 3.0.5
and provide them also - or- compile these programs by yourself. ;-)

I'd say this is rather a bug ticket for the authors of 
https://aur.archlinux.org/packages/wxgtk3-dev
https://aur.archlinux.org/packages/wxgtk3-dev-opt
than a bug issue of mmex or wxWidget.

**Comment by moritzhoewer** [comment]
Sorry to pitch into your discussion, but I am having a deja-vu from about 2 months ago when pretty much this same issue was already discussed.

The way I see it, there are people like @PMaff who are fine with building mmex and wxWidgets from source - and there are people like @MartinX3 who want to use a package manager (actually @MartinX3, you are maintaining the AUR package for mmex, aren't you?)

Problem is that for the packages, there is a clash of philosophies. wxWidgets should have released a "proper" version to go into the oficial repos, but they didn't. One could install the dev version into `/usr/local`, which is where it would go for people like @PMaff who compile from source, but @MartinX3 can't/won't do that for the package, as it would break conventions on where packages are supposed to put files...so there seems to be a dead end.

One possible solution to both of those problems was discussed above (https://github.com/moneymanagerex/moneymanagerex/issues/4255#issuecomment-1094258429): putting the wxWidgets dev version into `/opt/`. But there were some issues (https://github.com/moneymanagerex/moneymanagerex/issues/4255#issuecomment-1094869624). The latest update on that (https://github.com/moneymanagerex/moneymanagerex/issues/4255#issuecomment-1109624569) still had no solution, but I don't really get it. It really should work.

**Comment by PMaff** [comment]
@moritzhoewer : the main problem is that the package providers of 
https://aur.archlinux.org/packages/wxgtk3-dev
https://aur.archlinux.org/packages/wxgtk3-dev-opt
use a location for the non offical version 3.1.6, where other programs/packages expect official version 3.0.5 .

That will never work. Neither in Arch nor in any other Linux distribution.
The only way to go with a unofficial version like 3.1.6 is to deploy it in `/usr/local` so that the other packages
which expect 3.0.5 in the official place /usr/lib/, can also find it there.

**Comment by MartinX3** [comment]
The `-opt` package installs the wxWidgets the lib in `/opt/`.

In the end it's the fault of the wxWidget developers.
So many companies behind it, still their release cycle is so slow and their release policy so bad.

**Comment by PMaff** [comment]
The wxWidget guys say that 3.1.5/3.1.6 is not released now.
So it is not their fault.

The release cycle is what it is, if other applications want to use that unofficial version, then the package providers
have to think, how it can be installed without destroying the basis for those who do not want to install the latest and greatest
of wxWidget.
No one forces the various programs to use 3.1.6 or even 3.1.7 .

**Comment by MartinX3** [comment]
It's a bad habit to use a non stable dependency and expect others to always support it.

**Comment by PMaff** [comment]
The non stable 3.1.6  was anounced by wxWidget as non released.
Anyone who wants to use non released software has to take the consequences.
If I am taking part in a beta program of -say- Windows 11, then I also have to take care what I am doing.
I am not forced to use that beta version, I decide to take the latest and greatest.

If the makers of 
https://aur.archlinux.org/packages/wxgtk3-dev
https://aur.archlinux.org/packages/wxgtk3-dev-opt
then install the 3.1.6 somewhere, where others expect 3.0.5 then there is something wrong with
the installation process of these packages.

You @MartinX3  just trusted that these guys of wxgtk3-dev and wxgtk3-dev-opt did the right thing,
which was not the case obviously.

**Comment by moritzhoewer** [comment]
I think @MartinX3 is referring to mmex using a non-stable dependency of wxWidgets

**Comment by MartinX3** [comment]
Not really.
If you do custom self compile hacks to workaround this problem, it's fine.

But we package users expect it to work by not package guideline violating procedures of installing software.

Also mmex ignores the custom path of wxwidgets I set with the compiling parameters.

So it's a bug of mmex.

**Comment by PMaff** [comment]
mmex is just fine with /usr/local/lib/wx .
It works on Manjaro, Suse, Ubuntu.

If the makers of the packages 
https://aur.archlinux.org/packages/wxgtk3-dev
https://aur.archlinux.org/packages/wxgtk3-dev-opt
install it in a place, where other applications are not expecting 3.0.5 then everything would be just fine.

**Comment by MartinX3** [comment]
Which precompiled application using the standard wx 3.0.5 packages expects it at `/opt/` where the `-opt` package installed it?
Wtf

**Comment by PMaff** [comment]
Btw. self compiling is not "hacking".
It's just normal and the guys who build 
https://aur.archlinux.org/packages/wxgtk3-dev
https://aur.archlinux.org/packages/wxgtk3-dev-opt
are also using https://github.com/wxWidgets/wxWidgets/releases/download/v3.1.6/wxWidgets-3.1.6.tar.bz2

**Comment by MartinX3** [comment]
The average user doesn't do the hacky way of manually self compiling.


Yes, aur packages are either binary packages or source code packages.
You install them from the package manager and do no it professional stuff by yourself.

**Comment by PMaff** [comment]
Then one should tell the makers of
https://aur.archlinux.org/packages/wxgtk3-dev
https://aur.archlinux.org/packages/wxgtk3-dev-opt
that they are putting a self compiled hacky version 3.1.6 into a place where the official packages
are expecting the official non-hacky version 3.0.5 .
;-)

**Comment by MartinX3** [comment]
I don't know if you need sleep, but the `-opt` package puts it into a custom path.

**Comment by PMaff** [comment]
I am not using -opt, I just use the way described here:
https://wiki.wxwidgets.org/Compiling_and_getting_started
which installs it into `/usr/local/lib` .
This approach works on Manjaro, Ubuntu and Suse.

**Comment by MartinX3** [comment]
Letting a package installs software at this path is a violation.

https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s09.html

```
The /usr/local hierarchy is for use by the system administrator when installing software locally. It needs to be safe from being overwritten when the system software is updated. It may be used for programs and data that are shareable amongst a group of hosts, but not found in /usr.

Locally installed software must be placed within /usr/local rather than /usr unless it is being installed to replace or upgrade software in /usr.
```

**Comment by PMaff** [comment]
" unless it is being installed to replace or upgrade software in /usr" ...
The question is, if a non-official 3.1.6 should replace a official 3.0.5 without all the other applications knowing about this.

**Comment by MartinX3** [comment]
It's a package.
Packages get updated by the package manager.
Packages are not allowed to use `/use/local`.

Packages with conflicting paths excluding each other by the conflict meta data in their packages.

**Comment by PMaff** [comment]
> It's a package. Packages get updated by the package manager. Packages are not allowed to use `/usr/local`.
> 
> Packages with conflicting paths excluding each other by the conflict meta data in their packages.

It's a package, that is not released and it should not be installed into a place, where other packages are expecting a released 3.0.5 of wxWidget. This is the reason why all the other programs fail afterwards.
The place in /usr/local is ok as long as it is your own machine and as long as you do "sudo" anyway.
If you are not admin/root of the machine, you should not try to install wxWidget 3.1.5 or 3.1.6 anyway,
then you should ask the admin of that machine anyway.

**Comment by PMaff** [comment]
If one does not like to install into `/usr/local`  or if he cannot (since he cannot do `sudo`) to avoid the clash with wxWidget 3.0.5 and other programs, all that remains is, that he cannot use the latest of wxWidget in 3.1.5,3.1.6 or 3.1.7 and therefore also has to use an older version of mmex.
`¯\_(ツ)_/¯`

**Comment by MartinX3** [comment]
I don't know why you always want to discuss nonsense.

`/usr/local` is no place for packages.
AUR packages are packages.

We are users, not developers or some kind of weird people who need to manually compile software on their own computers.

The only solution is to fix the mmex source code to allow custom wxWidget paths at compile time.

Or move away from wxWidgets which has a bad project management.

**Comment by PMaff** [comment]
If "/usr/local is no place for packages", then you cannot use unreleased software like wxWidget 3.1.5 , 3.1.6 as replacement for 3.0.5, when other packages rely on 3.0.5 .
That simple.

This is NOT a problem of mmex or wxWidget.
wxWidget  writes clearly about it's state.
mmex states clearly that the newer versions are not using wxWidget 3.0.5 .
That is all correct.

If you cannot install into `/usr/local` then you have to use older version 3.0.5 of wxWidget and the corresponding version
of mmex and all is fine. 
You just cannot use the latest and greatest then.

You have to wait then, until someone corrects the install path of 3.1.6 or compiles all the other programs expecting 3.0.5 in that
place with 3.1.6 in that place.
That's not "weird" or "nonsense" - it is just a matter of dependencies, that are not resolved in ARCH when installing 3.1.6 wrongly into the place of 3.0.5 .

**Comment by PMaff** [comment]
To be honest, I would close this issue since it is an issue of the package creators of
https://aur.archlinux.org/packages/wxgtk3-dev
https://aur.archlinux.org/packages/wxgtk3-dev-opt
and the package creators of programs that rely on wxWidget 3.0.5 .

@whalley @vomikan : your opinion about closing this?

**Comment by moritzhoewer** [comment]
I agree that the discussion is going in circles. However I would not close this yet.

I was able to reproduce the issue Martin had (compiling MMEX with a custom wxWidget path and then installing it, with the result that MMEX won't start unless I manually specify the custom library location via the `LD_LIBRARY_PATH` environment variable.

Unfortunately I have not managed to finduch time to look into this further, so I can't say if MMEX is at fault here or the Arch Linux packaging or the packaging script. So maybe we can keep the ticket open for now and I will try to look into this some more soon

**Comment by PMaff** [comment]
`ldconfig` is mandatory after installing wxWidget 3.1.6 or 3.1.5 into `/usr/local `.
I do not know if the AUR package mechanisms are doing that.
See above
https://github.com/moneymanagerex/moneymanagerex/issues/4255#issuecomment-1133758293

**Comment by MartinX3** [comment]
I don't know why you cycle over and over again, ignoring everything which could be wrong with wxwidgets or mmex and just blame random package managers.
That's childish.

There is a flag to choose the wxwidgets path at compile time and this flag is ignored by mmex or something in the mmex project. 

And `/usr/local` is not allowed to be used by packages from a package management system.
There a standards and standards shouldn't get violated

**Comment by PMaff** [comment]
I have to repeat endlessly, since you do not see the point: 
if the package creators of 
https://aur.archlinux.org/packages/wxgtk3-dev
https://aur.archlinux.org/packages/wxgtk3-dev-opt
install a version 3.1.5 or 3.1.6 into a directory, where other packages expect a 3.0.5
then this has to fail and the fail is not the fault of wxWidget or mmex.
I am sorry for you, that you rely on these falsely installed packages.

**Comment by MartinX3** [comment]
I can run software using 3.0.5 just fine?

**Comment by moritzhoewer** [comment]
Ok, so I got it to work with wxWidgets installed in `/opt/`

You can instruct the linker to bake the full path to the shared libraries into the final executable. There are two ways to do that. One is to modify the CMake, but I am guessing that would probably be complicated to get right for all the different systems the CMake file has to work on. So I chose the second option, which is to tell the linker via an environment variable `LD_RUN_PATH`.
```
export LD_RUN_PATH="/location/one:/location/two"
```
(see section 3 of https://homepages.inf.ed.ac.uk/imurray2/compnotes/library_linking.txt - all the way at the bottom)

Now my MMEX finds all the libraries, even though they are in `/opt/`
```
ldd /usr/bin/mmex
	linux-vdso.so.1 (0x00007ffcadfd9000)
	libcurl.so.4 => /usr/lib/libcurl.so.4 (0x00007fb10d136000)
	libwx_gtk3u_core-3.1.so.6 => /opt/wxgtk-dev/lib/libwx_gtk3u_core-3.1.so.6 (0x00007fb10c87b000)
	libwx_gtk3u_qa-3.1.so.6 => /opt/wxgtk-dev/lib/libwx_gtk3u_qa-3.1.so.6 (0x00007fb10c849000)
	libwx_gtk3u_html-3.1.so.6 => /opt/wxgtk-dev/lib/libwx_gtk3u_html-3.1.so.6 (0x00007fb10c767000)
	libwx_baseu_xml-3.1.so.6 => /opt/wxgtk-dev/lib/libwx_baseu_xml-3.1.so.6 (0x00007fb10c752000)
	libwx_gtk3u_aui-3.1.so.6 => /opt/wxgtk-dev/lib/libwx_gtk3u_aui-3.1.so.6 (0x00007fb10c6aa000)
	libwx_gtk3u_stc-3.1.so.6 => /opt/wxgtk-dev/lib/libwx_gtk3u_stc-3.1.so.6 (0x00007fb10c41c000)
	libwx_gtk3u_webview-3.1.so.6 => /opt/wxgtk-dev/lib/libwx_gtk3u_webview-3.1.so.6 (0x00007fb10c3ea000)
	libwx_baseu-3.1.so.6 => /opt/wxgtk-dev/lib/libwx_baseu-3.1.so.6 (0x00007fb10c0c8000)
	liblua5.3.so.5.3 => /usr/lib/liblua5.3.so.5.3 (0x00007fb10c08a000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007fb10be53000)
        [...]
```

MMEX can't really be blamed here, as the output from `wx-config-gtk3` does make its way into the linker flags properly. But unfortunately, that only ensures that the libraries are found at link time and not at runtime. Changing that requires adding a few more linker flags, which as I wrote above is probably complicated to get right. 

I would therefore suggest we modify the AUR script - @MartinX3 if you are ok with switching the dependency to `wxgtk3-dev-opt` then I can send you a patch or write a comment on the AUR page or whatever.

**Comment by MartinX3** [comment]
@moritzhoewer
Thank you, that's a helpful post.

I don't know if it's a good way if packages randomly modify the environment variables.
What exactly do you want to modify in the AUR script?
The CMake?

Yes, I'm fine with `wxgtk3-dev-opt` as a workaround.
That's the reason I listed it. :)

**Comment by moritzhoewer** [comment]
The environment variable is only needed while building. It's not required later when the package is installed.

Basically, I changed
```
cmake -DCMAKE_BUILD_TYPE=Release -DwxWidgets_CONFIG_EXECUTABLE=/opt/wxgtk-dev/bin/wx-config-gtk3 .
                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
because that's where `wxgtk3-dev-opt` installs it.

and then added
```
export LD_RUN_PATH=/opt/wxgtk-dev/lib
```
before the
```
cmake --build . --target package
```
which makes the linker hardcode the library path

While I was at it, I also added a
```
export CXXFLAGS=-w
```
right at the top, to suppress all the compiler warnings that spam the console when building MMEX ;)


So my full `build()` function in the `PKGBUILD` looks like this now:
```
build() {
  cd "${srcdir}"/moneymanagerex

  # Disable all warnings when building, then configure CMake
  export CXXFLAGS=-w
  cmake -DCMAKE_BUILD_TYPE=Release -DwxWidgets_CONFIG_EXECUTABLE=/opt/wxgtk-dev/bin/wx-config-gtk3 .

  # Make sure the linker bakes the non-standard path for wxWidgets into the executable
  export LD_RUN_PATH=/opt/wxgtk-dev/lib
  cmake --build . --target package
}
```
and then of course I change the dependency up at the top.

**Comment by moritzhoewer** [comment]
About the `FindwxWidgets.cmake` thing - I don't know, I haven't looked into the MMEX build toolchain to be honest. Maybe it's possible to adapt the CMake files in MMEX so they detect non-standard locations of wxWidgets and bake them in. Maybe it's even easy, then we should maybe do that. But if it's hard or complicated or breaks on Windows/Max then I am kind of with @PMaff here. If we decide to setup our system with libraries - albeit experimental versions of libraries - in some random non-standard location, we can't really blame MMEX for not handling that as we intend.

**Comment by MartinX3** [comment]
The problem is that mmex decided to use a non stable dependency version of wxWidgets which isn't used by default in the linux distributions since it's not the stable version.

And the normal aur wxwidgets package installs the experimental wxWidgets version in the standard path.
Just the aur `-opt` package puts it into a custom location.

If your solution works then future users won't be afraid to use mmex if the use software depending on the of course preferred stable version of wxWidgets (3.0.5 at the moment).

**Comment by MartinX3** [comment]
Thank you very much.
That works and the mmex package uses now a custom location which doesn't conflict with the stable wxWidgets version every linux distribution uses by default.

**Comment by MartinX3** [comment]
FYI
3.1.7 should get released soon, which is the last version before the stable 3.2.0 release.
https://github.com/wxWidgets/wxWidgets/milestone/12

With 3.2.0 finally a new stable wxWidget version will be deployed in the standard distribution repos.

**Comment by gombosg** [comment]
I'm trying to package mmex for Fedora but also run into the same issue: wxSqlite package is linked against wxGTK 3.0, but this software needs 3.1 in order to compile. I'm thinking about how to proceed - packaging requires using package dependencies where possible.

**Comment by PMaff** [comment]
https://github.com/moneymanagerex/moneymanagerex/issues/2824
